core/dbt/include/global_project/macros/materializations/models/clone/clone.sql

```diff
@@ -0,0 +1,68 @@
+{%- materialization clone, default -%}
+
+  {%- set relations = {'relations': []} -%}
+
+  {%- if not state_relation -%}
+      -- nothing to do
+      {{ log("No relation found in state manifest for " ~ model.unique_id, info=True) }}
+      {{ return(relations) }}
+  {%- endif -%}
+
+  {%- set existing_relation = load_cached_relation(this) -%}
+
+  {%- if existing_relation and not flags.FULL_REFRESH -%}
+      -- noop!
+      {{ log("Relation " ~ existing_relation ~ " already exists", info=True) }}
+      {{ return(relations) }}
+  {%- endif -%}
+
+  {%- set other_existing_relation = load_cached_relation(state_relation) -%}
+
+  -- If this is a database that can do zero-copy cloning of tables, and the other relation is a table, then this will be a table
+  -- Otherwise, this will be a view
+
+  {% set can_clone_table = can_clone_table() %}
+
+  {%- if other_existing_relation and other_existing_relation.type == 'table' and can_clone_table -%}
+
+      {%- set target_relation = this.incorporate(type='table') -%}
+      {% if existing_relation is not none and not existing_relation.is_table %}
+        {{ log("Dropping relation " ~ existing_relation ~ " because it is of type " ~ existing_relation.type) }}
+        {{ drop_relation_if_exists(existing_relation) }}
+      {% endif %}
+
+      -- as a general rule, data platforms that can clone tables can also do atomic 'create or replace'
+      {% call statement('main') %}
+          {{ create_or_replace_clone(target_relation, state_relation) }}
+      {% endcall %}
+
+      {% set should_revoke = should_revoke(existing_relation, full_refresh_mode=True) %}
+      {% do apply_grants(target_relation, grant_config, should_revoke=should_revoke) %}
+      {% do persist_docs(target_relation, model) %}
+
+      {{ return({'relations': [target_relation]}) }}
+
+  {%- else -%}
+
+      {%- set target_relation = this.incorporate(type='view') -%}
+
+      -- TODO: this should probably be illegal
+      -- I'm just doing it out of convenience to reuse the 'view' materialization logic
+      {%- do context.update({
+          'sql': get_clone_target(state_relation),
+          'compiled_code': get_clone_target(state_relation)
+      }) -%}
+
+      -- reuse the view materialization
+      -- TODO: support actual dispatch for materialization macros
+      {% set search_name = "materialization_view_" ~ adapter.type() %}
+      {% if not search_name in context %}
+          {% set search_name = "materialization_view_default" %}
+      {% endif %}
+      {% set materialization_macro = context[search_name] %}
+      {% set relations = materialization_macro() %}
+      {{ return(relations) }}
```

Comment:
This is definitely the bit I felt worst about in https://github.com/dbt-labs/dbt-core/pull/7258.

I opened an issue to track real dispatch support for materialization macros: https://github.com/dbt-labs/dbt-core/issues/7799. Until we do that, this will work _much_ of the time, but it won't support adapters that inherit from other adapters. E.g. `dbt-postgres` &rarr; `dbt-redshift`, ergo Redshift is allowed to use `postgres__` implementations.

As far as hard-overwriting the context values for `sql` + `compiled_code`: Gross, right? The other potential approach I can think of is, making materialization macros accept `kwargs`, so that `sql`/`compiled_code` can be passed in explicitly rather than set + pulled out of the ~aether~ context.

Reply:
Good breakdown of the questions @ChenyuLInx!

Thinking through the logic we'd want from the `view` materialization:
- Handling the cases where the object already exists as a table, and needs to be dropped first
- Knowing whether this adapter wants to `create or replace view` versus create temp + alter/rename/swap
- Applying grants, persisting descriptions as comments - I think we still might actually want these things!

So - I would argue that's the right way to go. I'm definitely interested in a world where materializations are more composable than they are currently, so that more-complex build strategies ("clone an object from one env into another") can call into simpler ones ("make a view").

Macros are already pretty composable. Materializations are just macros, but they pull all their inputs from this node's g-d context, rather than receiving them as explicit arguments.